source/fab/metrics.py

```diff
@@ -0,0 +1,214 @@
+
+stop
+    closes pipes & process
+    return metrics from summary pipe
+
+"""
+import datetime
+import json
+import logging
+from collections import defaultdict
+from multiprocessing import Process, Pipe
+from multiprocessing.connection import Connection
+from pathlib import Path
+from typing import Optional, Dict
+
+logger = logging.getLogger('fab')
+
+# the pipe for individual metrics
+_metric_recv_conn: Optional[Connection] = None
+_metric_send_conn: Optional[Connection] = None
+
+# the pipe connection for the final collated metrics
+_collated_metrics_recv_conn: Optional[Connection] = None
+_collated_metrics_send_conn: Optional[Connection] = None
+
+# the process which receives individual metrics
+_metric_recv_process: Optional[Process] = None
+
+
+def init_metrics():
+    """
+    Create the pipe for sending metrics, the process to read them, and another pipe to push the final collated data.
+
+    Only one call to init_metrics can be called before calling stop_metrics.
+
+    """
+    global _metric_recv_conn, _metric_send_conn
+    global _collated_metrics_recv_conn, _collated_metrics_send_conn
+    global _metric_recv_process
+
+    if any([_metric_recv_conn, _metric_send_conn, _collated_metrics_recv_conn, _collated_metrics_send_conn]):
+        raise ConnectionError('Metrics already initialised. Only one concurrent user of init_metrics is expected.')
+
+    # the pipe connections for individual metrics
+    _metric_recv_conn, _metric_send_conn = Pipe(duplex=False)
+
+    # the pipe connections for the collated metrics, once they've stopped being sent
+    _collated_metrics_recv_conn, _collated_metrics_send_conn = Pipe(duplex=False)
+
+    # start the receiving process
+    _metric_recv_process = Process(
+        target=_read_metric,
+        daemon=True,  # todo: test this thoroughly, manually
+    )
+    _metric_recv_process.start()
+
+
+def _read_metric():
+    """
+    Intended to run as a child process, reading metrics created by other child processes.
+
+    Reads from the metric pipe until the pipe is closed,
+    at which point it pushes the collated metrics onto the "collated metrics" pipe and ends.
+
+    """
+    # An example metric is the time taken to preprocess a file; metrics['preprocess c']['my_file.c']
+    metrics: Dict[str, Dict[str, float]] = defaultdict(dict)
+
+    # todo: can we do this better?
+    # we run in a subprocess, so we get a copy of _metric_send_conn before it closes.
+    # when the calling process finally does close it, we'll still have an open copy of it,
+    # so the connection will still be considered open!
+    # therefore we close *OUR* copy of it now.
+    _metric_send_conn.close()
+
+    logger.debug('read_metric: waiting for metrics')
+    num_recorded = 0
+    while True:
+        try:
+            metric = _metric_recv_conn.recv()
+        except EOFError:
+            break
+
+        group, name, value = metric
+        metrics[group][name] = value
```

Comment:
Obviously there is scope here to over run available memory during execution. Future development should look to buffering and spooling out blocks of records.

Reply:
Agreed (but unlikely?). Again, a third party tool would cover this so we'd like to postpone but we've also added a _todo_ in the code.